File: atcoder/abc399/C_Make_it_Forest.py
import sys
from collections import defaultdict

sys.setrecursionlimit(10**6)
input = sys.stdin.readline


def max_acyclic_subgraph(graph, n):
    visited = [False] * (n + 1)
    in_stack = [False] * (n + 1)
    cycle_edges = 0

    def dfs(node, parent):
        nonlocal cycle_edges
        visited[node] = True
        in_stack[node] = True

        for neighbor in graph.get(node, []):
            if not visited[neighbor]:
                if dfs(neighbor, node):
                    pass
            elif in_stack[neighbor] and neighbor != parent:
                cycle_edges += 1

        in_stack[node] = False
        return False

    for node in graph:
        if not visited[node]:
            dfs(node, -1)

    return cycle_edges


n, m = map(int, input().split())
e = defaultdict(list)
for _ in range(m):
    u, v = map(int, input().split())
    e[u].append(v)
    e[v].append(u)
print(max_acyclic_subgraph(e, n))

File: atcoder/abc399/C_Make_it_Forest.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <cstring>

using namespace std;

const int MAX_N = 200011;  // Adjust based on constraints

vector<bool> visited(MAX_N, false);
vector<bool> in_stack(MAX_N, false);
int cycle_edges = 0;

unordered_map<int, vector<int>> graph;

void dfs(int node, int parent) {
    visited[node] = true;
    in_stack[node] = true;

    for (int neighbor : graph[node]) {
        if (!visited[neighbor]) {
            dfs(neighbor, node);
        } else if (in_stack[neighbor] && neighbor != parent) {
            cycle_edges++;
        }
    }

    in_stack[node] = false;
}

int max_acyclic_subgraph(int n) {
    cycle_edges = 0;
    fill(visited.begin(), visited.end(), false);
    fill(in_stack.begin(), in_stack.end(), false);

    for (auto& [node, _] : graph) {
        if (!visited[node]) {
            dfs(node, -1);
        }
    }

    return cycle_edges;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    cout << max_acyclic_subgraph(n) << endl;

    return 0;
}